app/src/main/java/mozilla/lockbox/view/EditItemFragment.kt

```diff
@@ -49,48 +82,152 @@ class EditItemFragment : BackableFragment(), EditItemDetailView {
-        closeKeyboard()
+    private fun setKeyboardFocus(view: View) {
+        val focusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            if (!hasFocus) {
+                closeKeyboard()
+            }
+        }
+
+        view.inputHostname.onFocusChangeListener = focusChangeListener
+        view.inputUsername.onFocusChangeListener = focusChangeListener
+
+        view.inputPassword.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            togglePasswordVisibility.accept(Unit)
+            if (!hasFocus) {
+                closeKeyboard()
+            }
+        }
     }
 
-    override var isPasswordVisible: Boolean = false
-        set(value) {
-            assertOnUiThread()
-            field = value
-            updatePasswordVisibility(value)
+    private fun setTextWatcher(view: View) {
+        val textInputs = listOf<Pair<TextInputEditText, TextInputLayout>>(
+            Pair(view.inputHostname, view.inputLayoutHostname),
+            Pair(view.inputUsername, view.inputLayoutUsername),
+            Pair(view.inputPassword, view.inputLayoutPassword)
+        )
+
+        for (input in textInputs) {
+            input.first.addTextChangedListener(
+                buildTextWatcher(input.second)
+            )
         }
+    }
 
-    override val togglePasswordClicks: Observable<Unit>
-        get() = view!!.btnPasswordToggle.clicks()
+    private fun buildTextWatcher(errorLayout: TextInputLayout): TextWatcher {
+        return object : TextWatcher {
+            override fun beforeTextChanged(
+                charSequence: CharSequence,
+                start: Int,
+                count: Int,
+                after: Int
+            ) {
+            }
 
-    override val learnMoreClicks: Observable<Unit>
-        get() = view!!.learnMore.clicks()
+            override fun onTextChanged(
+                charSequence: CharSequence,
+                start: Int,
+                before: Int,
+                count: Int
+            ) {
+            }
 
-    override val deleteClicks: Observable<Unit>
-        get() = view!!.deleteEntryButton.clicks()
+            override fun afterTextChanged(editable: Editable) {
+                val inputText: String? = errorLayout.editText?.text.toString()
 
-    override val closeEntryClicks: Observable<Unit>
-        get() = view!!.toolbar.navigationClicks()
+                when (errorLayout.id) {
+                    R.id.inputLayoutHostname -> {
+                        handleHostnameChanges(errorLayout, inputText)
+                    }
+                    R.id.inputLayoutUsername -> {
+                        handleUsernameChanges(errorLayout, inputText)
+                    }
+                    R.id.inputLayoutPassword -> {
+                        handlePasswordChanges(errorLayout, inputText)
+                    }
+                    else -> {
+                        errorLayout.error = null
+                    }
```

Comment:
Unnit: This is the case I was talking about: using `switch`/`when` especially in multiple places is an indicator that you should be using a State pattern, so a class for each state. In this case, this would map to a `TextWatcher` class per input.

The trade offs for this aren't obvious, especially on constrained devices with low method count limits.